mssql_python/parameter_helper.py

```diff
@@ -0,0 +1,220 @@
+
+    Uses simple character scanning approach - does NOT parse SQL contexts
+    (strings, comments, identifiers). This means %(name)s patterns inside SQL
+    string literals or comments WILL be detected as parameters.
+
+    Args:
+        sql: SQL query string with %(name)s placeholders
+
+    Returns:
+        List of parameter names in order of appearance (with duplicates if reused)
+
+    Examples:
+        >>> parse_pyformat_params("SELECT * FROM users WHERE id = %(id)s")
+        ['id']
+
+        >>> parse_pyformat_params("WHERE name = %(name)s OR email = %(name)s")
+        ['name', 'name']
+
+        >>> parse_pyformat_params("SELECT * FROM %(table)s WHERE id = %(id)s")
+        ['table', 'id']
+    """
+    params = []
+    i = 0
+    length = len(sql)
+
+    while i < length:
+        # Look for %(
+        if i + 2 < length and sql[i] == "%" and sql[i + 1] == "(":
+            # Find the closing )
+            j = i + 2
+            while j < length and sql[j] != ")":
+                j += 1
+
+            # Check if we found ) and it's followed by 's'
+            if j < length and sql[j] == ")":
+                if j + 1 < length and sql[j + 1] == "s":
+                    # Extract parameter name
+                    param_name = sql[i + 2 : j]
+                    params.append(param_name)
+                    i = j + 2
+                    continue
+
+        i += 1
+
+    return params
+
+
+def convert_pyformat_to_qmark(sql: str, param_dict: Dict[str, Any]) -> Tuple[str, Tuple[Any, ...]]:
+    """
+    Convert pyformat-style query to qmark-style for ODBC execution.
+
+    Validates that all required parameters are present and builds a positional
+    parameter tuple. Supports parameter reuse (same parameter appearing multiple times).
+
+    Args:
+        sql: SQL query with %(name)s placeholders
+        param_dict: Dictionary of parameter values
+
+    Returns:
+        Tuple of (rewritten_sql_with_?, positional_params_tuple)
+
+    Raises:
+        KeyError: If required parameter is missing from param_dict
+
+    Examples:
+        >>> convert_pyformat_to_qmark(
+        ...     "SELECT * FROM users WHERE id = %(id)s",
+        ...     {"id": 42}
+        ... )
+        ("SELECT * FROM users WHERE id = ?", (42,))
+
+        >>> convert_pyformat_to_qmark(
+        ...     "WHERE name = %(name)s OR email = %(name)s",
+        ...     {"name": "alice"}
+        ... )
+        ("WHERE name = ? OR email = ?", ("alice", "alice"))
+    """
+    # Support %% escaping - replace %% with a placeholder before parsing
+    # This allows users to have literal % in their SQL
+    escaped_sql = sql.replace("%%", "\x00ESCAPED_PERCENT\x00")
+
+    # Extract parameter names in order
+    param_names = parse_pyformat_params(escaped_sql)
+
+    if not param_names:
+        # No parameters found - restore escaped %% and return as-is
+        restored_sql = escaped_sql.replace("\x00ESCAPED_PERCENT\x00", "%")
```

Comment:
use the same marker here as well
```suggestion
        restored_sql = escaped_sql.replace(_ESCAPE_MARKER, "%")
```